nixos/modules/services/databases/postgresql.nix

```diff
@@ -108,77 +114,35 @@ in
-          This option will never delete existing databases, especially not when the value of this
-          option is changed. This means that databases created once through this option or
-          otherwise have to be removed manually.
-        '';
-        example = [
-          "gitea"
-          "nextcloud"
-        ];
-      };
-
-      ensureUsers = mkOption {
-        type = types.listOf (types.submodule {
-          options = {
-            name = mkOption {
-              type = types.str;
-              description = ''
-                Name of the user to ensure.
-              '';
-            };
-            ensurePermissions = mkOption {
-              type = types.attrsOf types.str;
-              default = {};
-              description = ''
-                Permissions to ensure for the user, specified as an attribute set.
-                The attribute names specify the database and tables to grant the permissions for.
-                The attribute values specify the permissions to grant. You may specify one or
-                multiple comma-separated SQL privileges here.
-
-                For more information on how to specify the target
-                and on which privileges exist, see the
-                <link xlink:href="https://www.postgresql.org/docs/current/sql-grant.html">GRANT syntax</link>.
-                The attributes are used as <code>GRANT ''${attrName} ON ''${attrValue}</code>.
-              '';
-              example = literalExample ''
-                {
-                  "DATABASE nextcloud" = "ALL PRIVILEGES";
-                  "ALL TABLES IN SCHEMA public" = "ALL PRIVILEGES";
-                }
-              '';
-            };
-          };
-        });
-        default = [];
+      statements = mkOption {
+        type = types.lines;
+        default = "";
         description = ''
-          Ensures that the specified users exist and have at least the ensured permissions.
-          The PostgreSQL users will be identified using peer authentication. This authenticates the Unix user with the
-          same name only, and that without the need for a password.
-          This option will never delete existing users or remove permissions, especially not when the value of this
-          option is changed. This means that users created and permissions assigned once through this option or
-          otherwise have to be removed manually.
+          <emphasis>Idempotent</emphasis> SQL statements to be executed by PostgreSQL. Useful for:
+          <itemizedlist>
+            <listitem><para>creating databases</para></listitem>
+            <listitem><para>creating local accounts with socket authentication</para></listitem>
+            <listitem><para>granting certain permissions on a database</para></listitem>
+          </itemizedlist>
+          <para>
+            See https://stackoverflow.com/a/18389184 for suggestions on writing idempotent sql
+            statements for PostgreSQL.
```

Comment:
Please correct me if I'm wrong, but this wouldn't fix issues like changed collation requirements for databases (as it happened with `matrix-synapse`), right?

Or in other words, what would happen, if
* the collation of a database has to be changed
* the module-author of an arbitrary service modifies the sql statement in `services.postgresql.statements`
* a user deploys this configuration to a server with an existing database

?

Reply:
The intention is not to fix the situation you described, but instead push volatile db work off of nixos and onto the sysadmin.

To answer your question: release notes explaining to users what they need to do should be issued. A special mention of backups should also be mentioned.